### ISSUE TITLE
Macros with Arguments

### ISSUE THREAD

**Issue by devd** [tabatkins/bikeshed#812]
One thing that is really useful in Latex at least was macros with arguments. So, for example,

```  
\newcommand{\wbalsup}[1] { This is the Wikibook about LaTeX  supported by #1}
\wbalsup{Wikipedia}
```
(from https://en.wikibooks.org/wiki/LaTeX/Macros)

works well. Currently, iiuc, Bikeshed only supports macros without any arguments. Arguments would help us a lot in the suborigin spec. We could define the suborigin serialization format in the macro and have it centralized in one space. I imagine something simple that (say) only allows up to 1-2 arguments at the most would be more that enough but would be valuable to a lot of specs.

**Comment by tabatkins** [comment]
Hmm, interesting. I allow file inclusions with arguments currently; would that suit your needs?

(/me goes to check if file inclusions are properly documented at all, gets frustrated again, decides to finally start working at porting all the documentation over to a Bikeshed document.)

**Comment by devd** [comment]
hmm .. that seems quite likely. Mind pointing me to the documentation/example? 😃

**Comment by tabatkins** [comment]
<https://github.com/tabatkins/bikeshed/blob/master/docs/markup.md#including-other-files>

**Comment by devd** [comment]
hmm .. yeah I guess that could work but putting a pre class=include everytime seems very painful; will that even work for something that needs to be inline?

**Comment by tabatkins** [comment]
It probably won't; I think I parse it with a fresh markdown context, so some bare text will end up being a paragraph.

I'll see what I can do about small inline replacements.

**Comment by tabatkins** [comment]
Okay, so design doc:

* The existing `Text Macros` metadata can provide the source here. It's got a decent grammar already. Currently they don't allow more macros in their content, but I'll fix that.
* Add a `<macro>` shorthand element. The macro name to invoke is written as a boolean argument (like the type of a DFN or A). It needs to be uppercase and with no value.
* Additional attributes on `<macro>` set up macros that are only valid within the replacement text (exactly like the `macros:` key in the include block's data).

Thus, `[FOO]` and `<macro FOO></macro>` are identical - both are inline-replaced with the contents of the FOO macro. But the latter form lets you do substitution into the macro replacement.

**Comment by devd** [comment]
that seems fine; but I wonder if a macro can take numbered arguments so that [FOO BAR BAZ] fills up with $1 = BAR $2=BAZ etc. My experience with Latex is that this slight ux improvement has huge value

**Comment by tabatkins** [comment]
Problem with that is that then you're limiting yourself to values without spaces, which is quite often not a reasonable limitation.

**Comment by devd** [comment]
yeah, you are right. but in that case people could still use the long form version above? Or we could find a way of allowing spaces? Latex, e.g., allows arguments in braces. In any case, I understand if you strongly prefer the approach you outlined: I don't have any history or context here and also lack your experience

**Comment by tabatkins** [comment]
Hmm, ok, I guess I can roll with that.  I'll need to move it to a new syntax, tho - the `[FOO]` syntax is a legacy mistake that I should have changed a long time ago, as it's far too easy to clash with innocent text.  The current restrictions to it being a single uppercase word make it not *too* bad, but allowing more things to match is unacceptable.

**Comment by tobie** [comment]
I'd love to use something like that to have a shorthand for ES descriptors which which is littered all over the WebIDL spec.

Something along the lines of

```xml
<macro DESCRIPTOR writable configurable />
```

which would expand to:

```html
<span class="descriptor">{ \[[Writable]]: <emu-val>true</emu-val>, \[[Enumerable]]: <emu-val>false</emu-val>, \[[Configurable]]: <emu-val>true</emu-val> }</span>
```

would be amazing. No idea if that's possible with the proposed syntax, though.
